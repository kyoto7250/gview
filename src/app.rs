use crate::{
    components::{
        commit_viewer::CommitViewer,
        content_viewer::ContentViewer,
        explorer::Explorer,
        operatable_components::{
            Message, MultipleTimesOperation, OnceOperation, OperatableComponent,
        },
    },
    repository::RepositoryInfo,
};
use crossterm::event::{self, Event, KeyCode, KeyEventKind};
use ratatui::{
    backend::CrosstermBackend,
    layout::{Constraint, Direction, Flex, Layout},
    terminal::Terminal,
    Frame,
};
use std::{
    io::{self, Stdout},
    iter::Once,
    ops::Mul,
    sync::{Arc, Mutex},
    time::{Duration, Instant},
};

// A simple alias for the terminal type used in this example.
pub type Tui = Terminal<CrosstermBackend<Stdout>>;

#[derive(Clone, Copy, Debug)]
enum FocusState {
    Filter,
    Filer,
    Commit,
    Viewer,
}

impl FocusState {
    fn next(self) -> FocusState {
        match self {
            FocusState::Filter => FocusState::Filer,
            FocusState::Filer => FocusState::Commit,
            FocusState::Commit => FocusState::Viewer,
            FocusState::Viewer => FocusState::Filter,
        }
    }
}

pub struct App {
    repository_info: Arc<Mutex<RepositoryInfo>>,
    left_main_chunk_percentage: u16,
    should_exit: bool,
    last_tick: Instant,
    focus_state: FocusState,
    explorer: Explorer,
    commit_viewer: CommitViewer,
    content_viewer: ContentViewer,
}

impl App {
    const TICK_RATE: Duration = Duration::from_millis(50);

    pub fn new(repository_info: RepositoryInfo) -> App {
        let repository = Arc::new(Mutex::new(repository_info));
        let mut app = Self {
            repository_info: Arc::clone(&repository),
            left_main_chunk_percentage: 15,
            should_exit: false,
            last_tick: Instant::now(),
            focus_state: FocusState::Filter,
            explorer: Explorer::new(),
            commit_viewer: CommitViewer::new(),
            content_viewer: ContentViewer::new(Arc::clone(&repository)),
        };
        app.handle_message(Message::MultipleTimes(MultipleTimesOperation::SetUp {
            repository: Arc::clone(&repository),
        }));
        app
    }

    fn process_focus(&mut self) {
        match self.focus_state {
            FocusState::Commit => self.commit_viewer.process_focus(),
            FocusState::Filter => self.explorer.filter.process_focus(),
            FocusState::Filer => self.explorer.filer.process_focus(),
            FocusState::Viewer => self.content_viewer.process_focus(),
        }
    }

    fn process_events(&mut self, code: KeyCode) -> Message {
        match self.focus_state {
            FocusState::Commit => self.commit_viewer.process_events(code),
            FocusState::Filter => self.explorer.filter.process_events(code),
            FocusState::Filer => self.explorer.filer.process_events(code),
            FocusState::Viewer => self.content_viewer.process_events(code),
        }
    }

    #[allow(unconditional_recursion)]
    fn handle_message(&mut self, message: Message) {
        // handle itself
        match message {
            Message::NoAction => return,
            Message::Once(OnceOperation::JumpToContentView) => {
                self.focus_state = FocusState::Viewer
            }
            Message::Once(OnceOperation::JumpToFiler) => self.focus_state = FocusState::Filer,
            _ => {}
        }

        let new_message = self.explorer.filer.handle_message(&message);
        self.handle_message(new_message);

        let new_message = self.explorer.filter.handle_message(&message);
        self.handle_message(new_message);

        let new_message = self.content_viewer.handle_message(&message);
        self.handle_message(new_message);

        let new_message = self.commit_viewer.handle_message(&message);
        self.handle_message(new_message);
    }

    pub fn run(&mut self, terminal: &mut Tui) -> io::Result<()> {
        while !self.should_exit {
            terminal.draw(|frame| {
                let _ = self.draw(frame);
            })?;
            self.handle_events()?;
            if self.last_tick.elapsed() >= Self::TICK_RATE {
                self.last_tick = Instant::now();
            }
        }
        Ok(())
    }

    fn handle_events(&mut self) -> io::Result<()> {
        let timeout = Self::TICK_RATE.saturating_sub(self.last_tick.elapsed());
        while event::poll(timeout)? {
            if let Event::Key(key) = event::read()? {
                if key.kind == KeyEventKind::Press {
                    match key.code {
                        KeyCode::Tab => {
                            self.process_focus();
                            self.focus_state = self.focus_state.next();
                            self.process_focus();
                        }
                        KeyCode::Char('q') => self.should_exit = true,
                        KeyCode::Char('<') => {
                            self.left_main_chunk_percentage =
                                self.left_main_chunk_percentage.saturating_sub(5).max(15);
                        }
                        KeyCode::Char('>') => {
                            self.left_main_chunk_percentage =
                                (self.left_main_chunk_percentage + 5).min(70);
                        }
                        _ => {
                            let message = self.process_events(key.code);
                            self.handle_message(message)
                        }
                    }
                }
            }
        }
        Ok(())
    }

    pub fn draw(&mut self, frame: &mut Frame) -> anyhow::Result<()> {
        let left_chunks = Layout::default()
            .direction(Direction::Horizontal)
            .constraints([
                Constraint::Percentage(self.left_main_chunk_percentage),
                Constraint::Percentage((100_u16).saturating_sub(self.left_main_chunk_percentage)),
            ])
            .split(frame.size());

        // chunks[0], chunks[1]
        let right_chunks = Layout::default()
            .direction(Direction::Vertical)
            .flex(Flex::Legacy)
            .constraints([Constraint::Length(3), Constraint::Length(5)].as_ref())
            .split(left_chunks[1]);

        let _ = self.explorer.draw(frame, left_chunks[0]);
        self.commit_viewer.draw(frame, right_chunks[0]);
        self.content_viewer.draw(frame, right_chunks[1]);
        Ok(())
    }
}
