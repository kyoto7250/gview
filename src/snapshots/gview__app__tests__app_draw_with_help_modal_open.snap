---
source: src/app.rs
expression: "format!(\"{:?}\", buffer)"
snapshot_kind: text
---
Buffer {
    area: Rect { x: 0, y: 0, width: 120, height: 40 },
    content: [
        "┌Partial Match───┐┌current commit (g: go to commit)────────────────────────────────────────────────────────────────────┐",
        "│                ││46ec93ec8aaddb99d1075cf8b11782ee83c6bd8e: Initial commit                                            │",
        "└────────────────┘└────────────────────────────────────────────────────────────────────────────────────────────────────┘",
        "┌1 files─────────┐┌test.txt────────────────────────────────────────────────────────────────────────────────────────────┐",
        "│>> test.txt╔ Key Configuration Help ══════════════════════════════════════════════════════════════════════╗           │",
        "│           ║Global Keys:                                                                                  ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║Tab           Switch focus between panels                                                     ║           │",
        "│           ║Ctrl+C        Exit gview                                                                      ║           │",
        "│           ║<             Decrease left panel width                                                       ║           │",
        "│           ║>             Increase left panel width                                                       ║           │",
        "│           ║?             Show this help modal                                                            ║           │",
        "│           ║ESC           Close help modal                                                                ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║Filter Panel:                                                                                 ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║Enter         Apply filter                                                                    ║           │",
        "│           ║Ctrl+A        Select all text                                                                 ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║File List Panel:                                                                              ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║↑/↓, j/k      Navigate files                                                                  ║           │",
        "│           ║Enter         Select file                                                                     ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║Commit Panel:                                                                                 ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║o             Open commit modal                                                               ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║Content Viewer:                                                                               ║           │",
        "│           ║                                                                                              ║           │",
        "│           ║↑/↓, j/k      Scroll content vertically                                                       ║           │",
        "│           ║←/→, h/l      Scroll content horizontally                                                     ║           │",
        "│           ║b             Toggle blame view                                                               ║           │",
        "│           ║n             Toggle line numbers                                                             ║           │",
        "│           ║g             Go to GitHub (if available)                                                     ║           │",
        "│           ╚══════════════════════════════════════════════════════════════════════════════════════════════╝           │",
        "│                ││                                                                                                    │",
        "│                ││                                                                                                    │",
        "│                ││                                                                                                    │",
        "└────────────────┘└────────────────────────────────────────────────────────────────────────────────────────────────────┘",
    ],
    styles: [
        x: 0, y: 0, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 0, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 19, y: 0, fg: DarkGray, bg: Reset, underline: Reset, modifier: BOLD,
        x: 51, y: 0, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 1, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 1, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 1, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 2, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 2, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 3, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 3, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 19, y: 3, fg: DarkGray, bg: Reset, underline: Reset, modifier: BOLD,
        x: 27, y: 3, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 4, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 4, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 4, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 5, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 5, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 5, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 5, fg: Cyan, bg: Reset, underline: Reset, modifier: BOLD,
        x: 25, y: 5, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 5, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 6, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 6, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 6, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 6, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 7, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 7, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 7, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 7, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 7, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 8, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 8, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 8, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 8, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 8, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 8, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 9, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 9, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 9, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 9, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 9, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 10, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 10, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 10, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 10, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 10, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 10, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 11, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 11, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 11, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 11, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 11, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 12, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 12, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 12, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 12, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 12, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 12, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 13, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 13, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 13, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 14, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 14, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 14, fg: Cyan, bg: Reset, underline: Reset, modifier: BOLD,
        x: 26, y: 14, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 14, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 15, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 15, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 15, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 15, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 16, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 16, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 16, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 16, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 16, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 17, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 17, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 17, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 17, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 17, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 17, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 18, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 18, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 18, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 19, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 19, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 19, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 19, fg: Cyan, bg: Reset, underline: Reset, modifier: BOLD,
        x: 29, y: 19, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 19, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 20, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 20, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 20, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 20, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 21, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 21, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 21, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 21, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 21, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 21, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 22, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 22, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 22, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 22, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 22, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 22, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 23, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 23, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 23, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 23, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 24, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 24, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 24, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 24, fg: Cyan, bg: Reset, underline: Reset, modifier: BOLD,
        x: 26, y: 24, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 24, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 25, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 25, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 25, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 25, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 26, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 26, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 26, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 26, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 26, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 26, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 27, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 27, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 27, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 27, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 28, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 28, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 28, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 28, fg: Cyan, bg: Reset, underline: Reset, modifier: BOLD,
        x: 28, y: 28, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 28, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 29, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 29, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 29, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 29, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 30, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 30, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 30, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 30, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 30, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 30, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 31, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 31, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 31, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 31, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 31, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 31, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 32, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 32, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 32, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 32, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 32, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 32, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 33, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 33, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 33, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 33, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 33, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 33, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 34, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 34, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 34, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 34, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 25, y: 34, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 34, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 35, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 35, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 35, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 108, y: 35, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 36, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 36, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 36, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 36, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 37, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 37, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 37, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 37, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 38, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 38, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 38, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 38, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 39, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 39, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
